File: src/main/java/composing/methods/Warehouse.java
package composing.methods;

import lombok.AllArgsConstructor;

import java.util.LinkedHashMap;

@AllArgsConstructor
public class Warehouse {
    private final int id;
    private final LinkedHashMap<Product, Integer> stock;

    public String generateStockReport() {
        StringBuilder report = new StringBuilder();
        report.append("Report for warehouse : " + id + "%n");

        stock.forEach((key, value) -> report.append("Product: " + key.getName() + " Price: " + key.getPrice() + " Stock : " + value + " units%n"));

        report.append("Total: " + stock.entrySet().stream().map(kvp -> kvp.getKey().getPrice() * kvp.getValue()).reduce(0.0, Double::sum) + "€");

        return report.toString();
    }
}

File: src/scala/src/main/scala/composing/methods/Warehouse.scala
package composing.methods

final case class Warehouse(
    id: Int,
    stock: Map[Product, Integer]
) {
  lazy val calculateStockValue: Double =
    stock.map { case (product, count) => product.price * count }.sum

  def generate(reportGenerator: Warehouse => String): String =
    reportGenerator(this)
}
